File: src/main/java/monopoly/model/cards/JSONCardFactory.java
package monopoly.model.cards;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import resources.json.JSONKey;

import java.util.ArrayList;
import java.util.Collections;

/**
 * JSONCardFactory: Et objekt til at hente alle start felter fra vores JSON resourcer
 *
 * @author Cecilie Krog Drejer, s185032
 * @author Joakim Bøegh Levorsen, s185023
 */

public class JSONCardFactory {
    /**
     * CreateChanceCards: Opretter Chancekort ud fra data fra JSON
     * 
     * @param JSONData Dataobjekt fra JSON
     * @param chanceStack Kortbunke, som kortene skal tilhøre
     * 
     * @author Cecilie Krog Drejer, s185032
     * @author Joakim Bøegh Levorsen, s185023
     * 
     * @return Returnerer et ArrayList af Cards
     * 
     * @throws JSONException
     */
    public static ArrayList<Card> createChanceCards(JSONObject JSONData, CardStack chanceStack) throws JSONException {
        JSONArray chanceCardData = JSONData.getJSONArray(JSONKey.CHANCE_CARDS.getKey());
        ArrayList<Card> chanceCards = new ArrayList<Card>();

        for (int i = 0; i < chanceCardData.length(); i++) {
            JSONObject chanceCard = chanceCardData.getJSONObject(i);
            int type = chanceCard.getInt(JSONKey.TYPE.getKey());
            switch (type) {
            case 0:
                chanceCards.add(GoToSpaceCard.create(chanceCard.getString(JSONKey.TEXT.getKey()),
                        chanceCard.getInt(JSONKey.SPACE.getKey())));
                break;
            case 1:
                chanceCards.add(GetLoseMoneyCard.create(chanceCard.getString(JSONKey.TEXT.getKey()),
                        chanceCard.getInt(JSONKey.AMOUNT.getKey())));
                break;
            case 2:
                chanceCards.add(GetOutOfJailCard.create(chanceCard.getString(JSONKey.TEXT.getKey())));
                break;
            case 3:
                String text = chanceCard.getString(JSONKey.TEXT.getKey());
                int spaceID = chanceCard.getInt(JSONKey.SPACE.getKey());
                GoToJailCard newCard = GoToJailCard.create(text, spaceID);
                chanceCards.add(newCard);
                break;
            default:
                throw new JSONException("Unexpected space type: " + type);
            }
        }
        Collections.shuffle(chanceCards);
        chanceStack.save();
        for (int i = 0; i < chanceCards.size(); i++) {
            Card card = chanceCards.get(i);
            card.setStackPosition(i);
            chanceStack.add(card);
            card.save();
        }
        return chanceCards;
    }

    /**
     * CreateCommunityChestCards: Opretter Prøv Lykken-kort ud fra data fra JSON
     * 
     * @param JSONData Dataobjekt fra JSON
     * @param communityChestStack Kortbunke, som kortene skal tilhøre
     * 
     * @author Cecilie Krog Drejer, s185032
     * @author Joakim Bøegh Levorsen, s185023
     * 
     * @return Returnerer et ArrayList af Cards
     * 
     * @throws JSONException
     */
    public static ArrayList<Card> createCommunityChestCards(JSONObject JSONData, CardStack communityChestStack)
            throws JSONException {
        JSONArray communityChestCardData = JSONData.getJSONArray(JSONKey.COMMUNITY_CHEST_CARDS.getKey());
        ArrayList<Card> communityChestCards = new ArrayList<Card>();

        for (int i = 0; i < communityChestCardData.length(); i++) {
            JSONObject communityChestCard = communityChestCardData.getJSONObject(i);
            int type = communityChestCard.getInt(JSONKey.TYPE.getKey());
            switch (type) {
            case 0:
                communityChestCards.add(GoToSpaceCard.create(communityChestCard.getString(JSONKey.TEXT.getKey()),
                        communityChestCard.getInt(JSONKey.SPACE.getKey())));
                break;
            case 1:
                communityChestCards.add(GetLoseMoneyCard.create(communityChestCard.getString(JSONKey.TEXT.getKey()),
                        communityChestCard.getInt(JSONKey.AMOUNT.getKey())));
                break;
            case 2:
                communityChestCards.add(GetOutOfJailCard.create(communityChestCard.getString(JSONKey.TEXT.getKey())));
                break;
            default:
                throw new JSONException("Unexpected space type: " + type);
            }
        }
        Collections.shuffle(communityChestCards);
        communityChestStack.save();
        for (int i = 0; i < communityChestCards.size(); i++) {
            Card card = communityChestCards.get(i);
            card.setStackPosition(i);
            communityChestStack.add(card);
            card.save();
        }
        return communityChestCards;
    }
}

File: src/main/java/monopoly/model/cards/JSONCardFactory.kt
package monopoly.model.cards

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import resources.json.JSONKey

import java.util.ArrayList
import java.util.Collections

/**
 * JSONCardFactory: Et objekt til at hente alle start felter fra vores JSON resourcer
 *
 * @author Cecilie Krog Drejer, s185032
 * @author Joakim Bøegh Levorsen, s185023
 */

object JSONCardFactory {
    /**
     * CreateChanceCards: Opretter Chancekort ud fra data fra JSON
     *
     * @param JSONData Dataobjekt fra JSON
     * @param chanceStack Kortbunke, som kortene skal tilhøre
     *
     * @author Cecilie Krog Drejer, s185032
     * @author Joakim Bøegh Levorsen, s185023
     *
     * @return Returnerer et ArrayList af Cards
     *
     * @throws JSONException
     */
    @Throws(JSONException::class)
    fun createChanceCards(JSONData: JSONObject, chanceStack: CardStack): ArrayList<Card> {
        val chanceCardData = JSONData.getJSONArray(JSONKey.CHANCE_CARDS.key)
        val chanceCards = ArrayList<Card>()

        for (i in 0 until chanceCardData.length()) {
            val chanceCard = chanceCardData.getJSONObject(i)
            val type = chanceCard.getInt(JSONKey.TYPE.key)
            when (type) {
                0 -> chanceCards.add(GoToSpaceCard.create(chanceCard.getString(JSONKey.TEXT.key),
                        chanceCard.getInt(JSONKey.SPACE.key)))
                1 -> chanceCards.add(GetLoseMoneyCard.create(chanceCard.getString(JSONKey.TEXT.key),
                        chanceCard.getInt(JSONKey.AMOUNT.key)))
                2 -> chanceCards.add(GetOutOfJailCard.create(chanceCard.getString(JSONKey.TEXT.key)))
                3 -> {
                    val text = chanceCard.getString(JSONKey.TEXT.key)
                    val spaceID = chanceCard.getInt(JSONKey.SPACE.key)
                    val newCard = GoToJailCard.create(text, spaceID)
                    chanceCards.add(newCard)
                }
                else -> throw JSONException("Unexpected space type: $type")
            }
        }
        Collections.shuffle(chanceCards)
        chanceStack.save()
        for (i in chanceCards.indices) {
            val card = chanceCards[i]
            card.stackPosition = i
            chanceStack.add(card)
            card.save()
        }
        return chanceCards
    }

    /**
     * CreateCommunityChestCards: Opretter Prøv Lykken-kort ud fra data fra JSON
     *
     * @param JSONData Dataobjekt fra JSON
     * @param communityChestStack Kortbunke, som kortene skal tilhøre
     *
     * @author Cecilie Krog Drejer, s185032
     * @author Joakim Bøegh Levorsen, s185023
     *
     * @return Returnerer et ArrayList af Cards
     *
     * @throws JSONException
     */
    @Throws(JSONException::class)
    fun createCommunityChestCards(JSONData: JSONObject, communityChestStack: CardStack): ArrayList<Card> {
        val communityChestCardData = JSONData.getJSONArray(JSONKey.COMMUNITY_CHEST_CARDS.key)
        val communityChestCards = ArrayList<Card>()

        for (i in 0 until communityChestCardData.length()) {
            val communityChestCard = communityChestCardData.getJSONObject(i)
            val type = communityChestCard.getInt(JSONKey.TYPE.key)
            when (type) {
                0 -> communityChestCards.add(GoToSpaceCard.create(communityChestCard.getString(JSONKey.TEXT.key),
                        communityChestCard.getInt(JSONKey.SPACE.key)))
                1 -> communityChestCards.add(GetLoseMoneyCard.create(communityChestCard.getString(JSONKey.TEXT.key),
                        communityChestCard.getInt(JSONKey.AMOUNT.key)))
                2 -> communityChestCards.add(GetOutOfJailCard.create(communityChestCard.getString(JSONKey.TEXT.key)))
                else -> throw JSONException("Unexpected space type: $type")
            }
        }
        Collections.shuffle(communityChestCards)
        communityChestStack.save()
        for (i in communityChestCards.indices) {
            val card = communityChestCards[i]
            card.stackPosition = i
            communityChestStack.add(card)
            card.save()
        }
        return communityChestCards
    }
}
